Feature/continuous integration deployment

Paddyjh opened the pull request.

Paddyjh committed:
Continuous-Integration-Deployment-initial-headings

Added the sub headings for Continuous Integration & Development

Paddyjh committed:
Continuous-Integration-Deployment-Subtopic-First-Draft

First draft of sub topic Automated Testing

Paddyjh committed:
Continuous-Integration-Deployment-continuous-integration-workflow

Added content for Continuous Integration Workflow subtopic.

The following topics have been discussed:

- Gitflow compliance

- Commiting practices

- CI platform integration

- Containerization Strategy

- Automated Testing Guidelines

- Failure Response Protocol

- Local and Manual Testing

Paddyjh committed:
Continuous-Integration-Deployment-deployment-process

Changed the title subtopic of "Deployment Strategies" to "Deployment Process".

I added information based on the following subheadings :

-Canary Strategy
-Automatic Deployment Process
-Feature Toggle
-Roll back Strategy

Paddyjh committed:
Continuous-Integration-Deployment-added-supporting-diagrams

I added supporting diagrams for topics where appropriate to make information easier to digest.

Isaac811 committed:
Continuous-Integration-Deployment-monitoring-feedback-and-security

Rosscrowley commented on Handbook/QA Handbook.md:
```diff
@@ -7,3 +7,86 @@
 ## Code Reviews
 
 ## Continuous Integration & Deployment
+
```
Hey @Paddyjh, I think we should maybe add a short introduction to the topic here.

Paddyjh replied:
Good idea, thanks for spotting. I'll add one now

Paddyjh replied:
Done

Rosscrowley commented on Handbook/QA Handbook.md:
Do we think these two bullet points could be broken down into more smaller bullet points?

Paddyjh replied:
I think that makes a lot of sense @Rosscrowley.

Also, @Isaac811, I feel like this could benefit from a more solid focus on security in continuous integration and deployment, specifically focusing on guidelines and best practices. This currently is a more general discussion on security practices

Isaac811 replied:
Please check the updated version and let me what do you think thx.

Paddyjh replied:
Hey @Isaac811, I think the new bullet point structure is looking good!

I still think this could benefit from discussing how these security practices are explicitly implemented in Continuous Integration and Deployment.

These two articles may be useful:

- https://cycode.com/blog/ci-cd-pipeline-security-best-practices/
- https://earthly.dev/blog/ci-cd-security-challenges-best-practices/  

If you have any questions, or if you think it is fine the way it is, just let me know!

Isaac811 replied:
I Will update the part. Thanks.

Paddyjh replied:
thanks @Isaac811 !
Give me a shout if you need a hand

Isaac811 replied:
Done，Thanks for the help.

Rosscrowley reviewed (changes requested):
I have left some thoughts that may need to be addressed

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -7,3 +7,86 @@
+### Monitoring and feedback loops
+![Monitoring_Feedback_Diagram](/Image_Folder/Monitoring_Feeback_Diagram.png)
+* View: During the project building process, keep update and send feedback to different parts is important. 
+* Adjustment: Feedback received by different departments to make changes. Improvement: Under the agile monitoring and feedback system, fast feedback, fast verification, and fast delivery are inevitable
```
I really like the diagram and information here, but I think this could be enhanced by incorporating a brief discussion on the benefits or impacts of implementing these feedback loops. Even just a few words at the end of each sentence could suffice.

Isaac811 replied:
Done.

Paddyjh reviewed (commented):
Left some suggestions for potential changes, let me know what you guys think and happy to answer any questions.

Rosscrowley commented on Handbook/QA Handbook.md:
```diff
@@ -7,3 +7,86 @@
+#### Gitflow compliance
+-	All developers must follow our company Gitflow branching strategy outlined in the diagram below and the following [link](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow).
+![gitflow-diagram](/Image_Folder/git-flow-diagram.png)
+#### Commiting practices
```
Spelling mistake here, "Commiting" should be "Committing". Thanks!

Paddyjh replied:
Done

Rosscrowley commented on Handbook/QA Handbook.md:
```diff
@@ -7,3 +7,86 @@
+
+#### Automatic Deployment Process
+- Deployment similar to integration should be automated using CI/CD tools like Jenkins 
+- Infrastructure as code should be used to ensure a consistent environment, tools like terraform and ansible should be utilised 
```
Terraform and Ansible should start with a capital letter. Thanks!

Paddyjh replied:
Done

Paddyjh committed:
Continuous-Integration-Deployment-introduction

Added introduction to address a comment on this branches current pull request

Isaac811 commented on Handbook/QA Handbook.md:
```diff
@@ -53,7 +53,27 @@
+- Developers should follow a Canary Deployment Strategy to ensure fast deployment and feedback while minimizing the risk of defects affecting users.
+	- A new version of the application should be rolled out to a select group of users 
+	- Data and feedback are gathered from the users and is incorporated into the new version before deploying to the entire user base
+	- If an issue is identified we can roll back the application to the previous state and address the issues, which will only affect the canary users
```
I think you could specify why canary is the best choice for our company specifically.

Paddyjh replied:
Good Idea, I just added there now. Thanks!

Paddyjh committed:
Continuous-Integration-Deployment-addressed-ross-comments

I have addressed the comments made by ross on the current request relating to my subtopics.

Isaac811 reviewed (changes requested):
Left 1 comment, let me know if you have questions.

Paddyjh committed:
Continuous-Integration-Deployment-added-justification-to-canary-strategy

I added a specific specification of why canary deployment is the best strategy for our team given its fast-moving nature but also the current prevalence of defects.

This addresses Isaac's comment on the current pull request.

Isaac811 committed:
continuous-integration-deloyment-updated

Isaac811 committed:
Merge branch 'feature/Continuous-Integration-Deployment' of https://github.com/Paddyjh/Software-Quality-Assurance-Group-10 into feature/Continuous-Integration-Deployment

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -7,3 +7,96 @@
+#### Feature Toggle
+- Developers should implement feature toggle (or flags) to enable the toggling on and off of features without the need to redeploy code
+- This is useful with canary as it allows us to make dynamic changes to our application without downtime.
+- A step-by-step approach to app release that splits traffic between a deployed version and a new version, releasing to a subset of users before a full rollout.
```
@Isaac811 Did you mean to add this section to the discussion on the feature toggle? What you have added seems to be describing canary releases which are already outlined above

Isaac811 replied:
Done.

Paddyjh replied:
Thanks for updating!

Isaac811 committed:
continues-integration-deployment-updated

Paddyjh committed:
Continuous-Integration-Deployment-reformatted-images-according-to-new-height-restriction

-I reformatted the images in this topic to align with our new height limit of 500px

-I also reformatted the section on Monitoring and feedback loops for better readability.

Paddyjh commented on Handbook/QA Handbook.md:
Hey @Isaac811 and @Rosscrowley, I updated all images in this pr to abide by the new 500px height limit

Isaac811 replied:
Looks good Thx!

Isaac811 committed:
Continuous-Integration-Deployment-updated-security-section

I updated the security section so that it was specific to CI/CD. This addresses a comment on the most recent pr

Isaac811 reviewed (approved):
Looks Good!

Paddyjh committed:
Merge branch 'develop' into feature/Continuous-Integration-Deployment

Rosscrowley reviewed (approved):
Looks good to me. Nice work guys!
